src/Microsoft.Restier.Core/ApiBuilderExtensions.cs

```diff
@@ -186,48 +183,91 @@ public static ApiBuilder ChainPrevious<T>(this ApiBuilder obj, Func<IServiceProv
+                    return factory(sp, next);
                 }
 
-                // TODO: See if in the future ActivatorUtilities could replace these implementation. For now its
-                // methods can't select constructor appropriately, base on avaialble services.
-                var ret = ChainedService.CreateViaReflection(typeof(TService), typeof(TImplement), sp, next);
-                var value = ret.Item1;
-                var creator = ret.Item2;
+                var instance = sp.GetService<TImplement>();
+                if (instance == null)
+                {
+                    return instance;
+                }
+
+                var nextProperty = typeof(TImplement).GetTypeInfo().GetProperties()
+                    .FirstOrDefault(e => e.SetMethod.IsPublic && e.PropertyType == typeof(TService));
+                if (nextProperty == null)
+                {
+                    factory = (serviceProvider, _) => serviceProvider.GetRequiredService<TImplement>();
+                    return instance;
+                }
 
-                // Create a factory proxy for the first call, to save the expression compilation cost to the 2nd
-                // call, in case the service is a singleton or for any reason only resolved once.
-                factory = (serviceProvider, nextService) =>
+                nextProperty.SetValue(instance, next());
+                factory = (serviceProvider, getNext) =>
                 {
-                    factory = ChainedService<TService>.CreateFactory(creator);
-                    creator = null;
-                    return factory(serviceProvider, nextService);
+                    var serviceProviderParam = Expression.Parameter(typeof(IServiceProvider));
+                    var nextParam = Expression.Parameter(typeof(Func<TService>));
+
+                    var value = Expression.Variable(typeof(TImplement));
+                    var getService = Expression.Call(
+                        typeof(ServiceProviderExtensions),
+                        "GetRequiredService",
+                        new[] { typeof(TImplement) },
+                        serviceProviderParam);
+                    var inject = Expression.Assign(
+                        Expression.MakeMemberAccess(value, nextProperty),
+                        Expression.Invoke(nextParam));
+
+                    var block = Expression.Block(
+                        typeof(TService),
+                        new[] { value },
+                        Expression.Assign(value, getService),
+                        inject,
+                        value);
+
+                    factory = LambdaExpression.Lambda<Func<IServiceProvider, Func<TService>, TService>>(
+                        block,
+                        serviceProviderParam,
+                        nextParam).Compile();
+                    nextProperty = null;
+                    return factory(serviceProvider, getNext);
```

Comment:
Then the question comes: can we directly compute the final hook object rather than build a lambda, compile and evaluate? Maybe it's not feasible, I am just guess.